utils/httpbody/httpbody.go

```diff
@@ -0,0 +1,42 @@
+package httpbody
+
+import (
+	"bytes"
+	"io"
+	"io/ioutil"
+	"net/http"
+)
+
+// BackupRequestBody returns a copy of the Request.Body, leaving it untouched
+// so the passed Request.Body will be still readable.
+func BackupRequestBody(req *http.Request) (io.ReadCloser, error) {
```

Comment:
why this function returns `io.ReadCloser` but the same one for response returns `[]byte`?

Reply:
we already talked about this.
They both do different things and have different purposes.
https://github.com/src-d/metadata-retrieval/pull/57#discussion_r335351660

Reply:
how come? They both read content, they both set a copy of the content to req/resp. And you use the absolutely the same as well, the comments use different words but also say the same. 
The only difference I see is this `[]byte`, `io.Reader` return type.

Reply:
This is related to what I wrote [here](https://github.com/src-d/metadata-retrieval/pull/57#discussion_r335454728). My idea was to keep the other function and remove this, in order to have the same thing for request and response as @smacker said.

Reply:
1. [line 49](https://github.com/src-d/metadata-retrieval/pull/57/files#diff-b99cf0f6f6715e2e68843770335d7f09R49) logs the content of the `Request.Body`, so it needes it.
1. since the `Request.Body` is needed at seen per `(1)`, it is read by [line 36](https://github.com/src-d/metadata-retrieval/pull/57/files#diff-b99cf0f6f6715e2e68843770335d7f09R36).
1. reading `(2)` the `Request.Body` can fail, so **if** it fails, it logs by [line 40](https://github.com/src-d/metadata-retrieval/pull/57/files#diff-b99cf0f6f6715e2e68843770335d7f09R40) `could not log the request: %s`.
1. then, the `if` at [line 37](https://github.com/src-d/metadata-retrieval/pull/57/files#diff-b99cf0f6f6715e2e68843770335d7f09R37) is needed to capture the failure when reading `2`.

Reply:
@smacker the `if` is still needed to capture the error that you're ignoring in your snippet at https://github.com/src-d/metadata-retrieval/pull/57#discussion_r335419273

Reply:
as it is taking forever can you just revert `logtransport.go` to the original state from master and remove `httpbody.go`? And the only other place where you read body is `readAPIErrorResponse`. You can replace with the same implementation as in logger in master:
```
func readAPIErrorResponse(resp *http.Response, errorResponse *apiErrorResponse) error {
	respBody, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return err
	}
	resp.Body = ioutil.NopCloser(bytes.NewBuffer(respBody))

	return json.Unmarshal(respBody, errorResponse)
}
```

You can send your suggestions in a separate PR.
`Close` can be also handled in separate PR.

Reply:
@dpordomingo I'm trying to understand, but I'm still having a hard time getting your point explanation. I've just tried this:
```
// logtransport.go

func (t *logTransport) RoundTrip(r *http.Request) (*http.Response, error) {
	t0 := time.Now()

	reqBody, err := httpbody.ReadRequestAndRestore(r)
	if err != nil {
		return nil, fmt.Errorf("could not backup the request before sending it: %s", err)
	}

	resp, err := t.T.RoundTrip(r)
	if err != nil {
		return nil, err
	}

	respBody, err := httpbody.ReadResponseAndRestore(resp)
	if err != nil {
		return nil, fmt.Errorf("could not backup the response before logging it: %s", err)
	}

	t.Logger.With(
		log.Fields{"elapsed": time.Since(t0), "response-code": resp.StatusCode, "url": r.URL, "request-body": string(reqBody), "response-body": string(respBody)},
	).Debugf("HTTP response")

	return resp, err
}
```

and:

```
// httpbody.go

func ReadResponseAndRestore(resp *http.Response) ([]byte, error) {
	reader, content, err := readAndGetNew(resp.Body)
	if err != nil {
		return nil, err
	}

	resp.Body = reader
	return content, nil
}

func ReadRequestAndRestore(req *http.Request) ([]byte, error) {
	reader, content, err := readAndGetNew(req.Body)
	if err != nil {
		return nil, err
	}

	req.Body = reader
	return content, nil
}
```

and it logs fine.

Reply:
I also tried adding closing of the response body as per [this comment](https://github.com/src-d/metadata-retrieval/pull/57#discussion_r3354156110 as follows:
```
func ReadResponseAndRestore(resp *http.Response) ([]byte, error) {
	body := resp.Body
	defer body.Close()

	reader, content, err := readAndGetNew(body)
	if err != nil {
		return nil, err
	}

	resp.Body = reader
	return content, nil
}
```

and it still works fine.

But as @smacker said, if it's easier to just revert I'm ok with it and postpone the improvement of closing the response body.

Reply:
done by 254ed27242e8745c89bbd34fd07e4eb09dab7160

Reply:
I still don't understand what was wrong with just adding the corresponding `ReadRequestAndRestore` since you already have implemented `ReadResponeAndRestore`. Also I tested it locally and it just seemed to work fine. But nvm, let's avoid dragging this further.